### ISSUE TITLE
SWIG depends on obsolete pcre library

### ISSUE THREAD

**Issue by Bluehorn** [swig/swig#2120]
Hi *,

according to https://pcre.org/, the pcre library version that SWIG is currently using is deprecated. Quote (emphasis added):

> current version, PCRE2, released in 2015, is now at version 10.39.
> The older, but still widely deployed PCRE library, originally released in 1997, is at version 8.45. This version of PCRE is now at end of life, and *is no longer being actively maintained*. Version 8.45 is expected to be the final release of the older PCRE library, and new projects should use PCRE2 instead.

This has been reported in Debian as https://bugs.debian.org/999948. Debian is planning to remove the old pcre from Debian for the bookworm release, but swig has a build dependency on it.

I intend to create a pull request to update the pcre dependency, but won't stop anybody else to work on this ;-)

Greetings, Torsten

**Comment by vadz** [comment]
It would be indeed nice to support PCRE2 and I did already think about doing it, but didn't have time to do it and I'm not sure when I'm going to have it, so if you can do it, please go ahead.

I don't know if there should be a choice between PCRE and PCRE2 at build time or if there is no need to maintain PCRE compatibility. Is/could there be any reason to not want to use PCRE2?

**Comment by Bluehorn** [comment]
@vadz Given that swig uses a really small subset of pcre features I'd say that we keep support for the old pcre for the time being as a fallback.
Otherwise this would have to be a major version bump for swig since it will not build everywhere where it built before. That could be a nasty surprise in some CI setups, not least the travis config. I guess that the travis environment still has the old pcre?

**Comment by Bluehorn** [comment]
Looking at the travis logs (now at https://app.travis-ci.com/github/swig/swig) it appears that at least on MacOS the pcre version is quite old:

``
PCRE Version: 8.44 2020-02-12

The command "./swig -version && ./swig -pcreversion" exited with 0.
```

I did not find a successful linux build with logs still available.

**Comment by vadz** [comment]
I'm usually the one arguing for preserving compatibility, so I'm definitely fine with keeping it if it's not too troublesome, TIA!

Concerning CI, SWIG uses GitHub Actions now instead of Travis and I think PCRE2 should be available there, but `apt-get install libpcre3-dev` in `Tools/CI-linux-install.sh` would need to be changed to use `libpcre2-dev` instead (upgrading to the next major version of the library means decreasing the version in the package name, but it is what it is...).

FWIW I think that the best way to isolate SWIG from system PCRE changes would be by including PCRE(2) as a submodule. `Tools/pcre-build.sh` relies on having a tarball in the current directory which is roughly equivalent but less convenient than already having them in the repository.

**Comment by ojwb** [comment]
I was aware of the `std::regex` deprecation situation, and I think that scuppers the idea.

Reading up on that a bit, I also saw it seems MSVC's implementation inherently has the equivalent of Perl's `/m` always enabled, whereas other implementations don't.  C++17 added a flag to enable it, but MSVC isn't implementing that.  This particular issue isn't relevant for SWIG (since we're dealing with symbol names which won't have newlines in) but it's not a good sign.

I'd argue against git submoduling or otherwise bundling a copy of pcre2 though.  Useful platforms will have a package available, and for other platforms downloading and building can be scripted.  Forcing everyone to download and use a particular version of pcre2 for the sake of platforms without a package manager wastes diskspace and CPU of developers, and exposes them to future security issues in the used version of PCRE (PCRE seems to average about 1 CVE a year).

**Comment by vadz** [comment]
Submodules have the advantage that you don't have to initialize them if you don't need them, so for me it's the best of both worlds: no need to use them when building under Linux, but everything you need is already in the repository if you need it under Windows.

I also don't think we could be impacted by any PCRE CVEs, we only use it with the programmer-provided regexes and while we do feed it user provided inputs, SWIG doesn't have any special permissions, so what could be the point of subverting it rather than just doing whatever the attacker needs to do directly?

**Comment by wsfulton** [comment]
Interesting research, especially about `std::regex`. I'd also argue against git submoduling or bundling a copy of PCRE2. Let's just swap out PCRE and replace it with PCRE2 and provide the same sort of support for using it as mentioned in configure.ac:

https://github.com/swig/swig/blob/9ffa15b3075dc2cf79e9425de0c8653680317e69/configure.ac#L79-L89

It's a successful recipe. The next release of SWIG (4.1.0) is one that can break compatibility, so changing the PCRE dependency is fine. The extra complexity of supporting the older PCRE as well as the newer PCRE2 doesn't seem to be warranted, so I wouldn't include end of life PCRE support as a fallback either. We can revisit this if many users report major problems that makes upgrading impossible. The pre-release testing that @jplesnik at Redhat does for us should give a good indicator of problems and there are fair number of people using master HEAD too who report difficulties.

@Bluehorn what are the timelines for this debian switch? It might sync up with our 4.1.0 release (which is long overdue, but I've got more time to work on getting this release out now).

**Comment by ojwb** [comment]
> I also don't think we could be impacted by any PCRE CVEs, we only use it with the programmer-provided regexes and while we do feed it user provided inputs, SWIG doesn't have any special permissions, so what could be the point of subverting it rather than just doing whatever the attacker needs to do directly?

I'm primarily thinking about running user-provided example interface files from bug reports through SWIG.  I try to look over them first (an additional reason I dislike non-reduced testcases), but there's potential for a remote attacker to execute code on my development machine by submitted a bug report with an interface file using a crafted regex to exploit an unpatched PCRE vulnerability.

> what are the timelines for this debian switch?

There are a lot of Debian packages which need updating for this: https://release.debian.org/transitions/html/pcre3-to-pcre2.html

I'd suggest we target this at 4.1.0 upstream, and once the changes are done @Bluehorn can cherry pick them if he needs a fixed in Debian before 4.1.0 is out.

I'd think the hard deadline for Debian here would be "before the release freeze for bookworm" which is probably ~ a year off, but ideally it'd be well before then so there's more time to address any problems from SWIG 4.1.0.

**Comment by Bluehorn** [comment]
Wow, I just had a few minutes to investigate using pcre2 in swig and just did a `git pull` in preparation. And I noticed that pcre2 support has already landed. Awesome!

```
torsten@pulsar:~/workspace/swig$ /usr/bin/swig -pcreversion
PCRE Version: 8.39 2016-06-14
torsten@pulsar:~/workspace/swig$ ./swig -pcreversion
PCRE2 Version: 10.39 2021-10-29
```
